File: app/src/main/java/coolpharaoh/tee/speicher/tea/timer/views/statistics/HorizontalBarGraph.java
package coolpharaoh.tee.speicher.tea.timer.views.statistics;

import android.app.Activity;

import androidx.core.content.ContextCompat;

import com.github.mikephil.charting.charts.HorizontalBarChart;
import com.github.mikephil.charting.components.AxisBase;
import com.github.mikephil.charting.components.Description;
import com.github.mikephil.charting.components.XAxis;
import com.github.mikephil.charting.components.YAxis;
import com.github.mikephil.charting.data.BarData;
import com.github.mikephil.charting.data.BarDataSet;
import com.github.mikephil.charting.data.BarEntry;
import com.github.mikephil.charting.formatter.ValueFormatter;

import java.util.ArrayList;
import java.util.List;

import coolpharaoh.tee.speicher.tea.timer.R;
import coolpharaoh.tee.speicher.tea.timer.core.tea.ColorConversation;
import coolpharaoh.tee.speicher.tea.timer.views.export_import.data_transform.pojo.StatisticsPOJO;

public class HorizontalBarGraph {
    public static final int VISIBLE_RANGE = 10;
    public static final float BAR_WIDTH = 0.8f;
    public static final float TEXT_SIZE = 16f;
    public static final int ANIMATION_DURATION = 2000;

    final HorizontalBarChart chart;
    final Activity activity;

    public HorizontalBarGraph(final HorizontalBarChart chart, final Activity activity) {
        this.chart = chart;
        this.activity = activity;

        configureGraph();
    }

    private void configureGraph() {
        configureChart();
        configureXAxis();
        configureYAxisRight();
        configureYAxisLeft();
    }

    private void configureChart() {
        final Description description = new Description();
        description.setText("");
        chart.setDescription(description);
        chart.getLegend().setEnabled(false);
        chart.setPinchZoom(false);
        chart.setScaleEnabled(false);
        chart.setDrawValueAboveBar(false);
    }

    private void configureXAxis() {
        final XAxis xAxis = chart.getXAxis();
        xAxis.setDrawGridLines(false);
        xAxis.setPosition(XAxis.XAxisPosition.BOTTOM);
        xAxis.setEnabled(true);
        xAxis.setDrawAxisLine(false);
        xAxis.setTextColor(ContextCompat.getColor(activity, R.color.text_black));
        xAxis.setTextSize(TEXT_SIZE);
    }

    private void configureYAxisRight() {
        final YAxis yRight = chart.getAxisRight();
        yRight.setDrawAxisLine(true);
        yRight.setDrawGridLines(false);
        yRight.setEnabled(false);
    }

    private void configureYAxisLeft() {
        final YAxis yLeft = chart.getAxisLeft();
        yLeft.setEnabled(false);
    }

    public void display(final List<StatisticsPOJO> statistics) {
        setMaximum(statistics);

        setLabels(statistics);

        setGraphData(statistics);

        setVisibleRange(statistics);

        chart.invalidate();
        chart.animateY(ANIMATION_DURATION);
    }

    private void setMaximum(final List<StatisticsPOJO> statistics) {
        final YAxis yLeft = chart.getAxisLeft();

        long max = 0;
        for (StatisticsPOJO statistic : statistics) {
            if (max <= statistic.counter) {
                max = statistic.counter;
            }
        }

        yLeft.setAxisMaximum(max);
        yLeft.setAxisMinimum(0f);
    }

    private void setLabels(final List<StatisticsPOJO> statistics) {
        final XAxis xAxis = chart.getXAxis();

        final ArrayList<String> values = new ArrayList<>();
        for (final StatisticsPOJO statistic : statistics) {
            final String teaName = truncateLongText(statistic.teaname, 5);
            values.add(teaName);
        }

        final int labelCount = Math.min(statistics.size(), VISIBLE_RANGE);
        xAxis.setLabelCount(labelCount);
        xAxis.setValueFormatter(new XAxisFormatter(values));
    }

    public String truncateLongText(final String text, final int maxLength) {
        if (text == null || text.isEmpty()) {
            return "";
        }
        return text.substring(0, Math.min(text.length(), maxLength));
    }

    private void setGraphData(final List<StatisticsPOJO> statistics) {
        final BarData data = new BarData(createBarDataSetFrom(statistics));

        data.setBarWidth(BAR_WIDTH);

        chart.setData(data);
    }

    private BarDataSet createBarDataSetFrom(final List<StatisticsPOJO> statistics) {
        final ArrayList<BarEntry> entries = createBarEntries(statistics);

        final BarDataSet barDataSet = new BarDataSet(entries, "Bar Data Set");
        setColors(statistics, barDataSet);

        barDataSet.setValueFormatter(new XAxisFormatter(null));
        barDataSet.setValueTextSize(TEXT_SIZE);

        return barDataSet;
    }

    private ArrayList<BarEntry> createBarEntries(final List<StatisticsPOJO> statistics) {
        final ArrayList<BarEntry> entries = new ArrayList<>();
        for (int i = 0; i < statistics.size(); i++) {
            entries.add(new BarEntry(i, statistics.get(i).counter));
        }
        return entries;
    }

    private void setColors(final List<StatisticsPOJO> statistics, final BarDataSet barDataSet) {
        final List<Integer> colors = setBarColors(statistics, barDataSet);

        setBarTextColors(statistics, barDataSet, colors);
    }

    private List<Integer> setBarColors(final List<StatisticsPOJO> statistics, final BarDataSet barDataSet) {
        final List<Integer> colors = new ArrayList<>();
        for (int i = 0; i < statistics.size(); i++) {
            colors.add(statistics.get(i).teacolor);
        }
        barDataSet.setColors(colors);
        return colors;
    }

    private void setBarTextColors(final List<StatisticsPOJO> statistics, final BarDataSet barDataSet, final List<Integer> colors) {
        final List<Integer> textColors = new ArrayList<>();
        for (int i = 0; i < statistics.size(); i++) {
            // because of a weird behavior I need to add the foreground color twice
            final int foregroundColor = ColorConversation.discoverForegroundColor(colors.get(i));
            textColors.add(foregroundColor);
            textColors.add(foregroundColor);
        }

        barDataSet.setValueTextColors(textColors);
    }

    private void setVisibleRange(List<StatisticsPOJO> statistics) {
        chart.setVisibleXRangeMaximum(VISIBLE_RANGE);
        chart.moveViewTo(0, statistics.size(), YAxis.AxisDependency.LEFT);
    }

    public void reset() {
        chart.fitScreen();
        chart.getData().clearValues();
        chart.getXAxis().setValueFormatter(null);
        chart.notifyDataSetChanged();
        chart.clear();
        chart.invalidate();
    }

    public static class XAxisFormatter extends ValueFormatter {

        private final List<String> values;

        public XAxisFormatter(final List<String> values) {
            this.values = values;
        }

        @Override
        public String getAxisLabel(final float value, final AxisBase axis) {
            if (value % 1 == 0) {
                return this.values.get((int) value);
            }
            return "";
        }

        @Override
        public String getFormattedValue(final float value) {
            return String.valueOf((int) Math.floor(value));
        }
    }
}

File: app/src/main/java/coolpharaoh/tee/speicher/tea/timer/views/statistics/HorizontalBarGraph.kt
package coolpharaoh.tee.speicher.tea.timer.views.statistics

import android.app.Activity
import androidx.core.content.ContextCompat
import com.github.mikephil.charting.charts.HorizontalBarChart
import com.github.mikephil.charting.components.AxisBase
import com.github.mikephil.charting.components.Description
import com.github.mikephil.charting.components.XAxis
import com.github.mikephil.charting.components.YAxis
import com.github.mikephil.charting.data.BarData
import com.github.mikephil.charting.data.BarDataSet
import com.github.mikephil.charting.data.BarEntry
import com.github.mikephil.charting.formatter.ValueFormatter
import coolpharaoh.tee.speicher.tea.timer.R
import coolpharaoh.tee.speicher.tea.timer.core.tea.ColorConversation.discoverForegroundColor
import coolpharaoh.tee.speicher.tea.timer.views.export_import.data_transform.pojo.StatisticsPOJO
import kotlin.math.floor

class HorizontalBarGraph(val chart: HorizontalBarChart, val activity: Activity) {
    init {
        configureGraph()
    }

    private fun configureGraph() {
        configureChart()
        configureXAxis()
        configureYAxisRight()
        configureYAxisLeft()
    }

    private fun configureChart() {
        val description = Description()
        description.text = ""
        chart.description = description
        chart.legend.isEnabled = false
        chart.setPinchZoom(false)
        chart.setScaleEnabled(false)
        chart.setDrawValueAboveBar(false)
    }

    private fun configureXAxis() {
        val xAxis = chart.xAxis
        xAxis.setDrawGridLines(false)
        xAxis.position = XAxis.XAxisPosition.BOTTOM
        xAxis.isEnabled = true
        xAxis.setDrawAxisLine(false)
        xAxis.textColor = ContextCompat.getColor(activity, R.color.text_black)
        xAxis.textSize = TEXT_SIZE
    }

    private fun configureYAxisRight() {
        val yRight = chart.axisRight
        yRight.setDrawAxisLine(true)
        yRight.setDrawGridLines(false)
        yRight.isEnabled = false
    }

    private fun configureYAxisLeft() {
        val yLeft = chart.axisLeft
        yLeft.isEnabled = false
    }

    fun display(statistics: List<StatisticsPOJO>) {
        setMaximum(statistics)

        setLabels(statistics)

        setGraphData(statistics)

        setVisibleRange(statistics)

        chart.invalidate()
        chart.animateY(ANIMATION_DURATION)
    }

    private fun setMaximum(statistics: List<StatisticsPOJO>) {
        val yLeft = chart.axisLeft

        var max: Long = 0
        for (statistic in statistics) {
            if (max <= statistic.counter) {
                max = statistic.counter
            }
        }

        yLeft.axisMaximum = max.toFloat()
        yLeft.axisMinimum = 0f
    }

    private fun setLabels(statistics: List<StatisticsPOJO>) {
        val xAxis = chart.xAxis

        val values = ArrayList<String>()
        for (statistic in statistics) {
            val teaName = truncateLongText(statistic.teaName, 5)
            values.add(teaName)
        }

        val labelCount = Math.min(statistics.size, VISIBLE_RANGE)
        xAxis.labelCount = labelCount
        xAxis.valueFormatter = XAxisFormatter(values)
    }

    fun truncateLongText(text: String?, maxLength: Int): String {
        return if (text.isNullOrEmpty()) {
            ""
        } else text.substring(0, Math.min(text.length, maxLength))
    }

    private fun setGraphData(statistics: List<StatisticsPOJO>) {
        val data = BarData(createBarDataSetFrom(statistics))

        data.barWidth = BAR_WIDTH

        chart.data = data
    }

    private fun createBarDataSetFrom(statistics: List<StatisticsPOJO>): BarDataSet {
        val entries = createBarEntries(statistics)

        val barDataSet = BarDataSet(entries, "Bar Data Set")
        setColors(statistics, barDataSet)

        barDataSet.valueFormatter = XAxisFormatter(null)
        barDataSet.valueTextSize = TEXT_SIZE

        return barDataSet
    }

    private fun createBarEntries(statistics: List<StatisticsPOJO>): ArrayList<BarEntry> {
        val entries = ArrayList<BarEntry>()
        for (i in statistics.indices) {
            entries.add(BarEntry(i.toFloat(), statistics[i].counter.toFloat()))
        }
        return entries
    }

    private fun setColors(statistics: List<StatisticsPOJO>, barDataSet: BarDataSet) {
        val colors = setBarColors(statistics, barDataSet)

        setBarTextColors(statistics, barDataSet, colors)
    }

    private fun setBarColors(statistics: List<StatisticsPOJO>, barDataSet: BarDataSet): List<Int> {
        val colors: MutableList<Int> = ArrayList()
        for (i in statistics.indices) {
            colors.add(statistics[i].teaColor)
        }
        barDataSet.colors = colors
        return colors
    }

    private fun setBarTextColors(statistics: List<StatisticsPOJO>, barDataSet: BarDataSet,
                                 colors: List<Int>) {
        val textColors: MutableList<Int> = ArrayList()
        for (i in statistics.indices) {
            // because of a weird behavior I need to add the foreground color twice
            val foregroundColor = discoverForegroundColor(colors[i])
            textColors.add(foregroundColor)
            textColors.add(foregroundColor)
        }

        barDataSet.setValueTextColors(textColors)
    }

    private fun setVisibleRange(statistics: List<StatisticsPOJO>) {
        chart.setVisibleXRangeMaximum(VISIBLE_RANGE.toFloat())
        chart.moveViewTo(0f, statistics.size.toFloat(), YAxis.AxisDependency.LEFT)
    }

    fun reset() {
        chart.fitScreen()
        chart.data.clearValues()
        chart.xAxis.valueFormatter = null
        chart.notifyDataSetChanged()
        chart.clear()
        chart.invalidate()
    }

    class XAxisFormatter(private val values: List<String>?) : ValueFormatter() {
        override fun getAxisLabel(value: Float, axis: AxisBase): String {
            return if (value % 1 == 0f) {
                values!![value.toInt()]
            } else ""
        }

        override fun getFormattedValue(value: Float): String {
            return floor(value.toDouble()).toInt().toString()
        }
    }

    companion object {
        const val VISIBLE_RANGE = 10
        const val BAR_WIDTH = 0.8f
        const val TEXT_SIZE = 16f
        const val ANIMATION_DURATION = 2000
    }
}